etc/migration/migrateUsersToShibboleth.js

```diff
@@ -0,0 +1,346 @@
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {String}     callback.userHash       The user hash object including the external id
+ * @api private
+ */
+function _findExternalId(userHash, callback) {
+    var userId = userHash.principalId;
+    Cassandra.runQuery('SELECT "loginId" FROM "AuthenticationUserLoginId" WHERE "userId" = ?', [userId], function(err, rows) {
+        if (err) {
+            log().error({'err': err}, 'Failed Cassandra query for user %s\'s loginId', userHash.displayName);
+            _writeErrorRow(userHash, 'Could not retrieve loginId for this user');
+            return callback(err);
+        }
+
+        var loginId = _.chain(rows)
+            .map(Cassandra.rowToHash)
+            .pluck('loginId')
+            .filter(function(loginId) {
+                // If user has more than one loginId, take the Google one
+                return loginId.split(':')[1] === 'google';
+            })
+            .first()
+            .value();
+
+        return callback(null, _.extend(userHash, {'loginId': loginId}));
+    });
+}
+
+/**
+ * Create new authentication records linked to Shibboleth EPPN for all users within the tenant
+ *
+ * @param  {Object[]}   allUsers                The array of all the users for a tenancy
+ * @param  {Function}   callback                Standard callback function
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {Object[]}   callback.mappedUsers    The Shibboleth users that were mapped to existing users
+ * @api private
+ */
+var _mapUsersToShibLogin = function(allUsers, callback) {
+    // Remove users with duplicate emails or invalid login IDs
+    var cleanedUsers = _.chain(allUsers)
+        .filter(function(user) {
+            var valid = user.loginId;
+            if (!valid) {
+                // We'll also end up here if they have a non-Google loginId
+                _writeErrorRow(user, 'This user has an invalid loginId');
+            }
+            return valid;
+        })
+        .compact()
+        .value();
+
+    if (_.isEmpty(cleanedUsers)) {
+        log().info('No suitable users found for tenant %s', tenantAlias);
+        return callback(null, null);
+    }
+
+    _createShibLoginRecords(cleanedUsers, function(err, mappedUsers) {
+        if (err) {
+            return callback(err);
+        }
+
+        return callback(null, mappedUsers);
+    });
+};
+
+/**
+ * Create new authentication records linked to Shibboleth EPPN for all users within the tenant
+ *
+ * @param  {Object[]}   users                   The users to map
+ * @param  {Function}   callback                Standard callback function
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {Object[]}   callback.mappedUsers    The Shibboleth users that were mapped to existing users
+ * @api private
+ */
+var _createShibLoginRecords = function(users, callback, _mappedUsers) {
```

Comment:
The use of variable naming around `users` v.s. `userHashes` is inconsistent making it a little tough to follow. Can you use `userHash` when referring to a raw object fetched from cassandra (which I believe is always the case in this file). The term `user` for a variable is typically used when we have a `User` object.